src/els_code_action_provider.erl

```diff
@@ -0,0 +1,183 @@
+     } = Range,
+   #{ title => Title
+    , command =>
+          els_utils:make_command( Title
+                                , <<"replace-lines">>
+                                , [#{ uri => Uri
+                                    , lines => Lines
+                                    , from => StartLine
+                                    , to => EndLine }])
+    }.
+
+-spec make_code_action(uri(), diagnostic()) -> [map()].
+make_code_action(Uri, #{ <<"message">> := Message
+                       , <<"range">> := Range } = _Diagnostic) ->
+    lager:info("make_code_action: [Message=~p]", [Message]),
+    CA1 = unused_variable_action(Uri, Range, Message),
+    CA2 = undefined_type_action(Uri, Range, Message),
+    CA1 ++ CA2.
+
+%%------------------------------------------------------------------------------
+
+-spec unused_variable_action(uri(), range(), binary()) -> [map()].
+unused_variable_action(Uri, Range, Message) ->
+    %% processing messages like "variable 'Foo' is unused"
+    %% TODO: this is a hot path, precompile the re's.
+    case re:run(Message, "variable '(.*)' is unused"
+               , [{capture, all_but_first, binary}]) of
+        {match, [UnusedVariable]} ->
+            make_unused_variable_action(Uri, Range, UnusedVariable);
+        _ -> []
+    end.
+
+-spec make_unused_variable_action(uri(), range(), binary()) -> [map()].
+make_unused_variable_action(Uri, Range, UnusedVariable) ->
+    #{ <<"start">> := #{ <<"character">> := _StartCol
+                       , <<"line">>      := StartLine }
+     , <<"end">> := _End
+     } = Range,
+    lager:info("make_unused_variable_action: [Range=~p]"
+              , [Range]),
+    %% processing messages like "variable 'Foo' is unused"
+    lager:info("make_unused_variable_action: [UnusedVariable=~p]"
+              , [UnusedVariable]),
+    Path = binary_to_list(els_uri:path(Uri)),
+    Lines = file2lines(Path),
+    Line = lists:nth(StartLine + 1, Lines),
+    lager:info("make_unused_variable_action: [Line=~p]"
+              , [Line]),
+
+    {ok, Tokens, _} = erl_scan:string(Line, 1, [return, text]),
+    lager:info("make_unused_variable_action: [Tokens=~p]"
+              , [Tokens]),
+    Replace = fun(Tok) ->
+                case Tok of
+                    {var, [{text, VarName}, _], _} -> "_" ++ VarName;
+                    {_,   [{text, Text   }, _], _} -> Text;
+                    {_,   [{text, Text   }, _]}    -> Text
+                end
+              end,
+    UpdatedLine = lists:flatten(lists:map(Replace, Tokens)) ++ "\n",
+    lager:info("make_unused_variable_action: [UpdatedLine=~p]"
+              , [UpdatedLine]),
+    [ replace_lines_action( Uri
+                      , <<"Add '_' to '", UnusedVariable/binary, "'">>
+                      , list_to_binary(UpdatedLine)
+                      , Range)].
+
+%%------------------------------------------------------------------------------
+
+%% Based on https://stackoverflow.com/a/2185533
```

Comment:
`els_text` contains utilities to split binaries into lines. There we use `Lines = binary:split(Text, <<"\n">>, [global]),` for the split, which should be good enough. We can expose that function and call it from here.